src/inmanta/plugins.py

```diff
@@ -313,37 +341,37 @@ def to_dsl_type(python_type: type[object]) -> inmanta_type.Type:
                 if len(args) == 1:
                     return inmanta_type.TypedDict(inmanta_type.Any())
 
-                return inmanta_type.TypedDict(to_dsl_type(args[1]))
+                return inmanta_type.TypedDict(to_dsl_type(args[1], location, resolver))
             else:
                 raise TypingException(None, f"invalid type {python_type}, dictionary types should be Mapping or dict")
 
         # List
         if issubclass(origin, Sequence):
-            if origin in [collections.abc.Sequence, list, typing.Sequence]:
+            if origin in [collections.abc.Sequence, list, Sequence]:
                 args = typing.get_args(python_type)
                 if not args:
                     return inmanta_type.List()
-                return inmanta_type.TypedList(to_dsl_type(args[0]))
+                return inmanta_type.TypedList(to_dsl_type(args[0], location, resolver))
             else:
                 raise TypingException(None, f"invalid type {python_type}, list types should be Sequence or list")
 
         # Set
         if issubclass(origin, collections.abc.Set):
             raise TypingException(None, f"invalid type {python_type}, set is not supported on the plugin boundary")
 
-    # TODO annotated types
-    # if typing.get_origin(t) is typing.Annotated:
-    #     args: Sequence[object] = typing.get_args(python_type)
-    #     inmanta_types: Sequence[plugin_typing.InmantaType] =
-    #     [arg if isinstance(arg, plugin_typing.InmantaType) for arg in args]
-    #     if inmanta_types:
-    #         if len(inmanta_types) > 1:
-    #             # TODO
-    #             raise Exception()
-    #         # TODO
-    #         return parse_dsl_type(inmanta_types[0].dsl_type)
-    #     # the annotation doesn't concern us => use base type
-    #     return to_dsl_type(args[0])
+        # Annotated
+        if origin is typing.Annotated:
+            for meta in python_type.__metadata__:  # type: ignore
+                if isinstance(meta, ModelType):
+                    if location is None or resolver is None:
+                        raise TypingException(
+                            None,
+                            f"Cannot parse {meta.model_type}, location or resolver were not provided to `to_dsl_type`",
+                        )
```

Comment:
Isn't this always a bug? Meaning, why do we even accept `None`?

Reply:
I changed it to allow None values mainly because of the dataclass PR. `DataclassExplainer` calls `to_dsl_type` and I was struggling to get the `location` and `namespace` there. And if I did get them from the `NamedType` they could still be None. And we only ever need this for Annotated types. So  in any other use case that we don't need them, we don't need to provide them.

Now that I'm writing (rubber-ducking this) I feel like we should indeed not accept None, my question is how do we guarantee that we always have the `namespace` and `location` set on `DataclassExplainer`?

Reply:
But we may have `Annotated` fields in the dataclasses, so we need to support that. The explainer seems to get a `problem` with an `entity` attached, so we can easily get what we need there. Once done, I would request @wouterdb's review for this file just in case.